README.md
# ModestImage


*Friendlier matplotlib interaction with large images*

ModestImage extends the matplotlib AxesImage class, and avoids
unnecessary calculation and memory when rendering large images (where most
image pixels aren't visible on the screen). It has the following
benefits over AxesImage:

 * Draw time is (roughly) independent of image size
 * Large ``numpy.memmap`` arrays can be visualized, without making an
   in-memory copy of the entire array. This enables visualization of
   images too large to fit in memory.

## Installation

```
pip install ModestImage
```
or
```
easy_install ModestImage
```

## Using ModestImage


The easiest way is to use the modified ``imshow`` function:

```
import matplotlib.pyplot as plt
from modest_image import ModestImage, imshow

ax = plt.gca()
imshow(ax, image_array, vmin=0, vmax=10)
plt.show()
```

``imshow`` accepts all the keyword arguments that the matplotlib
function does. The ``vmin`` and ``vmax`` keywords aren't necessary
but, if they are not provided, the entire image will be scanned to
determine the min/max values. This can be slow if the array is huge.

To create a ModestImage artist directly:

```
artist = ModestImage(data=array)
```

## Looking at very big FITS images


```
import matplotlib.pyplot as plt
import pyfits
from modest_image import imshow

ax = plt.gca()
huge_array = pyfits.open('file_name.fits', memmap=True)[0].data
artist = imshow(ax, huge_array, vmin=0, vmax=10)
plt.show()
```

This opens almost instantly, with a modest memory footprint.

## Why is Matplotlib Image Drawing Slow?


For the first draw request after setting the color mapping or data
array, AxesImage (the default matplotlib image class) calculates the
RGBA value for every pixel in the data array. That's a lot of work for
large images, and usually overkill given that the final rendering is
limited by screen resolution (usually 100K-1M pixels) and not image
resolution (often much more).

AxesImage compensates for this by saving the results of this
scaling. This means that subsequent renderings that only change the
position or zoom level are very fast. However, in interactive
situations where the data array or intensity scale change often,
AxesImage wastes lots of time calculating RGBA values for every pixel
in a (potentially large) data set. It also makes several temporary
arrays with size comparable to the original array, wasting memory.

## How is ModestImage faster?

ModestImage resamples the image array at each draw request, extracting
a smaller image whose resolution and extent are matched to the screen
resolution. Thus, the RGBA scaling step is much faster, since it takes
place only for pixels relevant for the current rendering.

This scheme does not take advantage of AxesImage's caching, and thus
redraws after move and zoom operations are slightly slower. However,
draws after colormap and data changes are substantially faster, and most
redraws are fast enough for interactive use.

## Performance and Tests

``speed_test.py`` compares the peformance of ModestImage and
AxesImage. For a 1000x1000 pixel image:

```
    Performace Tests for AxesImage

           time_draw: 186 ms per operation
           time_move: 19 ms per operation
      time_move_zoom: 28 ms per operation

    Performace Tests for ModestImage

          time_draw: 25 ms per operation
          time_move: 20 ms per operation
     time_move_zoom: 28 ms per operation
```

``time_draw`` is the render time after the cache has been cleared
(e.g. after ``set_data`` has been called, or the colormap has been
changed). ModestImage is slightly slower than, though still
competetive with, AxesImage for move and zoom operations where
AxesImage uses cached data.

Unit tests can be found in the ``tests`` directory. ModestImage does not
always produce results identical to AxesImage at the pixel level, due to
how it downsamples images. The discrepancy is minor, however, and disappears
if no downsampling takes place (i.e. a screen pixel samples <= 1 data pixel)


CHANGES.md
0.2 (unreleased)
----------------

- Fixed compatibility with Python 3 and Matplotlib 1.5.x

0.1 (2014-05-08)
----------------

- Initial release

tests/test.py
import pytest

from matplotlib import pyplot as plt
import matplotlib.image as mi

import numpy as np

from modest_image import ModestImage

x, y = np.mgrid[0:300, 0:300]
_data = np.sin(x / 10.) * np.cos(y / 30.)

def setup_function(func):
    plt.clf()
    plt.cla()

def default_data():
    return _data

def init(img_cls, data):
    fig = plt.figure()
    ax = fig.add_subplot(111)
    artist = img_cls(ax, data=data)
    ax.add_artist(artist)
    ax.set_aspect('equal')
    artist.norm.vmin = -1
    artist.norm.vmax = 1
    ax.set_xlim(0, data.shape[1])
    ax.set_ylim(0, data.shape[0])

    return artist

def check(label, modest, axes):
    """ Assert that images are identical, else save and fail """
    assert modest.figure is not axes.figure
    assert modest.figure is not axes.figure

    modest.figure.canvas.draw()
    axes.figure.canvas.draw()
    str1 = modest.figure.canvas.tostring_rgb()
    str2 = axes.figure.canvas.tostring_rgb()

    if str1 == str2:
        return

    result = 'PASS' if str1 == str2 else 'FAIL'
    modest_label = 'test_%s_modest' % label
    axes_label = 'test_%s_default' % label

    modest.axes.set_title(modest_label + " " + result)
    axes.axes.set_title(axes_label + " " + result)

    axes.figure.canvas.draw()
    modest.figure.canvas.draw()

    modest.figure.savefig(modest_label+'.pdf')
    axes.figure.savefig(axes_label+'.pdf')

    assert str1 == str2

def test_default():
    """ Zoomed out view """
    data = default_data()
    modest = init(ModestImage, data)
    axim = init(mi.AxesImage, data)
    check('default', modest.axes, axim.axes)

def test_move():
    """ move at default zoom """
    data = default_data()
    modest = init(ModestImage, data)
    axim = init(mi.AxesImage, data)
    xlim = modest.axes.get_xlim()
    delta = 50

    modest.axes.set_xlim(xlim[0] + delta, xlim[1] + delta)
    axim.axes.set_xlim(xlim[0] + delta, xlim[1] + delta)
    check('move', modest.axes, axim.axes)

def test_zoom():
    """ zoom in """
    data = default_data()
    modest = init(ModestImage, data)
    axim = init(mi.AxesImage, data)
    lohi = 200, 250
    modest.axes.set_xlim(lohi)
    axim.axes.set_xlim(lohi)
    modest.axes.set_ylim(lohi)
    axim.axes.set_ylim(lohi)

    check('zoom', modest.axes, axim.axes)

def test_zoom_out():
    """ zoom out """
    data = default_data()
    modest = init(ModestImage, data)
    axim = init(mi.AxesImage, data)
    lohi = -1000, 1000
    modest.axes.set_xlim(lohi)
    axim.axes.set_xlim(lohi)
    modest.axes.set_ylim(lohi)
    axim.axes.set_ylim(lohi)

    check('zoom_out', modest.axes, axim.axes)


INTRP_METHODS = ('nearest', 'bilinear', 'bicubic',
                 'spline16', 'spline36', 'hanning',
                 'hamming', 'hermite', 'kaiser',
                 'quadric', 'catrom', 'gaussian',
                 'bessel', 'mitchell', 'sinc', 'lanczos',
                 'none')

@pytest.mark.parametrize(('method'), INTRP_METHODS)
def test_interpolate(method):
    """ change interpolation """
    data = default_data()
    modest = init(ModestImage, data)
    axim = init(mi.AxesImage, data)

    lohi = 100, 150
    modest.axes.set_xlim(lohi)
    axim.axes.set_xlim(lohi)
    modest.axes.set_ylim(lohi)
    axim.axes.set_ylim(lohi)
    modest.set_interpolation(method)
    axim.set_interpolation(method)
    check('interp_%s' % method, modest.axes, axim.axes)

def test_scale():
    """change color scale"""

    data = default_data()
    modest = init(ModestImage, data)
    axim = init(mi.AxesImage, data)

    for im in [modest, axim]:
        im.norm.vmin = .7
        im.norm.vmin = .8

    check("cmap", modest.axes, axim.axes)

def test_unequal_limits():
    """Test different x/y scalings"""
    data = default_data()
    modest = init(ModestImage, data)
    axim = init(mi.AxesImage, data)

    for im in [modest, axim]:
        im.axes.set_aspect('auto')
        im.axes.set_xlim(20, 30)
        im.axes.set_ylim(10, 80)

    check('unequal_limits', modest.axes, axim.axes)

def test_alpha():
    """alpha changes """
    data = default_data()
    modest = init(ModestImage, data)
    axim = init(mi.AxesImage, data)

    for im in [modest, axim]:
        im.set_alpha(.3)

    check('alpha', modest.axes, axim.axes)


tests/test_speed.py
from time import time

from matplotlib import pyplot as plt
import matplotlib.image as mi

import numpy as np

from modest_image import ModestImage


x, y = np.mgrid[0:1000, 0:1000]
data = np.sin(x / 10.) * np.cos(y / 30.)


def setup(img_cls):

    plt.clf()
    plt.cla()
    ax = plt.gca()

    artist = img_cls(ax, data=data)

    ax.add_artist(artist)
    ax.set_aspect('equal')
    artist.norm.vmin = -1
    artist.norm.vmax = 1
    ax.set_xlim(0, 2000)
    ax.set_ylim(0, 2000)

    return artist, time()


def report(label, t0, t1, niter):
    print '%15s: %i ms per operation' % (label,
                                         (t1 - t0) * 1000 / niter)


def time_draw(img_cls, repeat=5):
    """render time for first draw after cache clear

    Simulates render after changing data or cmap
    """

    a, t = setup(img_cls)
    for i in range(repeat):
        a.changed()  # clear caches, simulate first render
        a.axes.figure.canvas.draw()
    report('time_draw', t, time(), repeat)


def time_move(img_cls, repeat=10):
    """ Render time for moving an image without cache clear """
    lims = [(0, 2000), (500, 2500)]
    a, t = setup(img_cls)
    ax = plt.gca()
    a.axes.figure.canvas.draw()
    t = time()
    for i in range(repeat):
        ax.set_xlim(lims[i % 2])
        ax.set_ylim(lims[i % 2])
        a.axes.figure.canvas.draw()
    report('time_move', t, time(), repeat)


def time_move_zoom(img_cls, repeat=10):
    """ Move image, at high zoom setting """
    delta = 100
    lims = [(0, 100), (500, 600)]
    a, t = setup(img_cls)
    ax = plt.gca()
    a.axes.figure.canvas.draw()
    t = time()
    for i in range(repeat):
        ax.set_xlim(lims[i % 2])
        ax.set_ylim(lims[i % 2])
        a.axes.figure.canvas.draw()
    report('time_move_zoom', t, time(), repeat)


def main():
    print 'Test image dimensions: %i x %i' % data.shape
    for im in [mi.AxesImage, ModestImage]:
        print '**********************************'
        print "Performace Tests for %s" % im.__name__
        print '**********************************'
        time_draw(im)
        time_move(im)
        time_move_zoom(im)


if __name__ == "__main__":
    main()


tests/test_imshow.py
from __future__ import print_function

import numpy as np
import matplotlib.pyplot as plt

from modest_image import ModestImage, imshow


def default_data():
    x, y = np.mgrid[0:100, 0:100]
    return np.sin(x / 25) * np.cos(y / 50)


def check_props(obj1, obj2, props):
    for p in props:
        v1 = obj1.__getattribute__('get_%s' % p)()
        v2 = obj2.__getattribute__('get_%s' % p)()
        assert v1 == v2


def check_artist_props(art1, art2):
    """Assert that properties of two artists are equal"""

    props = ['alpha', 'clim', 'clip_on', 'clip_path',
             'interpolation', 'label', 'rasterized',
             'resample', 'snap', 'url', 'visible', 'zorder']
    check_props(art1, art2, props)


def check_axes_props(ax1, ax2):
    props = ['aspect']
    check_props(ax1, ax2, props)


def setup_function(func):
    plt.clf()
    plt.cla()


def teardown_function(func):
    plt.close()


def test_imshow_creates_modest_image():
    """returns a modestImage"""
    data = default_data()
    fig = plt.figure()
    ax = fig.add_subplot(111)

    artist = imshow(ax, data)

    assert isinstance(artist, ModestImage)
    assert artist in artist.axes.images


def test_imshow_mimics_mpl():
    """properties of two axes and artists objects should be same"""

    data = default_data()
    fig = plt.figure()
    ax1 = fig.add_subplot(121)
    ax2 = fig.add_subplot(122)

    artist1 = ax1.imshow(data)
    artist2 = imshow(ax2, data)

    check_artist_props(artist1, artist2)
    check_axes_props(ax1, ax2)


modest_image/modest_image.py
from __future__ import print_function

import matplotlib
rcParams = matplotlib.rcParams

import matplotlib.image as mi
import matplotlib.colors as mcolors
import matplotlib.cbook as cbook
import numpy as np


class ModestImage(mi.AxesImage):

    """
    Computationally modest image class.

    ModestImage is an extension of the Matplotlib AxesImage class
    better suited for the interactive display of larger images. Before
    drawing, ModestImage resamples the data array based on the screen
    resolution and view window. This has very little affect on the
    appearance of the image, but can substantially cut down on
    computation since calculations of unresolved or clipped pixels
    are skipped.

    The interface of ModestImage is the same as AxesImage. However, it
    does not currently support setting the 'extent' property. There
    may also be weird coordinate warping operations for images that
    I'm not aware of. Don't expect those to work either.
    """

    def __init__(self, *args, **kwargs):
        if 'extent' in kwargs and kwargs['extent'] is not None:
            raise NotImplementedError("ModestImage does not support extents")

        self._full_res = None
        self._sx, self._sy = None, None
        self._bounds = None
        super(ModestImage, self).__init__(*args, **kwargs)

    def set_data(self, A):
        """
        Set the image array

        ACCEPTS: numpy/PIL Image A
        """
        self._full_res = A
        self._A = A

        if self._A.dtype != np.uint8 and not np.can_cast(self._A.dtype,
                                                         np.float):
            raise TypeError("Image data can not convert to float")

        if (self._A.ndim not in (2, 3) or
                (self._A.ndim == 3 and self._A.shape[-1] not in (3, 4))):
                raise TypeError("Invalid dimensions for image data")

        self._imcache = None
        self._rgbacache = None
        self._oldxslice = None
        self._oldyslice = None
        self._sx, self._sy = None, None

    def get_array(self):
        """Override to return the full-resolution array"""
        return self._full_res

    def _scale_to_res(self):
        """ Change self._A and _extent to render an image whose
        resolution is matched to the eventual rendering."""

        ax = self.axes
        shp = self._full_res.shape
        x0, x1, sx, y0, y1, sy = extract_matched_slices(ax, shp)
        # have we already calculated what we need?
        if (self._bounds is not None
            and sx >= self._sx and sy >= self._sy
            and x0 >= self._bounds[0] and x1 <= self._bounds[1]
            and y0 >= self._bounds[2] and y1 <= self._bounds[3]):
            return
        self._A = self._full_res[y0:y1:sy, x0:x1:sx]
        self._A = cbook.safe_masked_invalid(self._A)
        if self.origin == 'upper':
            self.set_extent([x0 - .5, x1 - .5, y1 - .5, y0 - .5])
        else:
            self.set_extent([x0 - .5, x1 - .5, y0 - .5, y1 - .5])
        self._sx = sx
        self._sy = sy
        self._bounds = (x0, x1, y0, y1)
        self.changed()

    def draw(self, renderer, *args, **kwargs):
        self._scale_to_res()
        super(ModestImage, self).draw(renderer, *args, **kwargs)


def main():
    from time import time
    import matplotlib.pyplot as plt
    x, y = np.mgrid[0:2000, 0:2000]
    data = np.sin(x / 10.) * np.cos(y / 30.)

    f = plt.figure()
    ax = f.add_subplot(111)

    # try switching between
    artist = ModestImage(ax, data=data)
    #artist = mi.AxesImage(ax, data=data)

    ax.set_aspect('equal')
    artist.norm.vmin = -1
    artist.norm.vmax = 1

    ax.add_artist(artist)
    ax.set_xlim(0, 1000)
    ax.set_ylim(0, 1000)

    t0 = time()
    plt.gcf().canvas.draw()
    t1 = time()

    print("Draw time for %s: %0.1f ms" % (artist.__class__.__name__,
                                          (t1 - t0) * 1000))

    plt.show()


def imshow(axes, X, cmap=None, norm=None, aspect=None,
           interpolation=None, alpha=None, vmin=None, vmax=None,
           origin=None, extent=None, shape=None, filternorm=1,
           filterrad=4.0, imlim=None, resample=None, url=None, **kwargs):
    """Similar to matplotlib's imshow command, but produces a ModestImage

    Unlike matplotlib version, must explicitly specify axes
    """

    if not axes._hold:
        axes.cla()
    if norm is not None:
        assert(isinstance(norm, mcolors.Normalize))
    if aspect is None:
        aspect = rcParams['image.aspect']
    axes.set_aspect(aspect)
    im = ModestImage(axes, cmap, norm, interpolation, origin, extent,
                     filternorm=filternorm,
                     filterrad=filterrad, resample=resample, **kwargs)

    im.set_data(X)
    im.set_alpha(alpha)
    axes._set_artist_props(im)

    if im.get_clip_path() is None:
        # image does not already have clipping set, clip to axes patch
        im.set_clip_path(axes.patch)

    # if norm is None and shape is None:
    #    im.set_clim(vmin, vmax)
    if vmin is not None or vmax is not None:
        im.set_clim(vmin, vmax)
    elif norm is None:
        im.autoscale_None()

    im.set_url(url)

    # update ax.dataLim, and, if autoscaling, set viewLim
    # to tightly fit the image, regardless of dataLim.
    im.set_extent(im.get_extent())

    axes.images.append(im)
    im._remove_method = lambda h: axes.images.remove(h)

    return im


def extract_matched_slices(ax, shape):
    """Determine the slice parameters to use, matched to the screen.

    :param ax: Axes object to query. It's extent and pixel size
               determine the slice parameters

    :param shape: Tuple of the full image shape to slice into. Upper
               boundaries for slices will be cropped to fit within
               this shape.

    :rtype: tulpe of x0, x1, sx, y0, y1, sy

    Indexing the full resolution array as array[y0:y1:sy, x0:x1:sx] returns
    a view well-matched to the axes' resolution and extent
    """
    ext = (ax.transAxes.transform([(1, 1)]) - ax.transAxes.transform([(0, 0)]))[0]
    xlim, ylim = ax.get_xlim(), ax.get_ylim()
    dx, dy = xlim[1] - xlim[0], ylim[1] - ylim[0]

    def _clip(val, hi):
        return int(max(min(val, hi), 0))

    y0 = _clip(min(ylim) - 5, shape[0])
    y1 = _clip(max(ylim) + 5, shape[0])
    x0 = _clip(min(xlim) - 5, shape[1])
    x1 = _clip(max(xlim) + 5, shape[1])

    sy = int(max(1, min((y1 - y0) / 5., np.ceil(abs(dy / ext[1])))))
    sx = int(max(1, min((x1 - x0) / 5., np.ceil(abs(dx / ext[0])))))

    return x0, x1, sx, y0, y1, sy

if __name__ == "__main__":
    main()
